google/cloud/pubsublite/internal/multipartition_publisher.cc

```diff
@@ -169,24 +164,43 @@ void MultipartitionPublisher::RouteAndPublish(PublishState state) {
       .then(OnPublish{std::move(state.publish_promise), partition});
 }
 
+void MultipartitionPublisher::TryPublishMessages() {
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (in_publish_loop_) return;
+    in_publish_loop_ = true;
+  }
+  while (true) {
+    std::deque<PublishState> messages;
+    uint32_t num_partitions;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (messages_.empty()) {
+        in_publish_loop_ = false;
+        return;
+      }
+      messages.swap(messages_);
+      num_partitions = static_cast<std::uint32_t>(partition_publishers_.size());
+    }
+    for (PublishState& state : messages) {
+      state.num_partitions = num_partitions;
+      RouteAndPublish(std::move(state));
+    }
+  }
+}
+
 future<StatusOr<MessageMetadata>> MultipartitionPublisher::Publish(
     PubSubMessage m) {
   PublishState state;
   state.message = std::move(m);
+  future<StatusOr<MessageMetadata>> to_return;
   {
     std::lock_guard<std::mutex> g{mu_};
-    // performing this under lock to guarantee that all of
-    // `initial_publish_buffer_` is flushed when publishers are created
-    if (partition_publishers_.empty()) {
-      initial_publish_buffer_.push_back(std::move(state));
-      return initial_publish_buffer_.back().publish_promise.get_future();
-    }
-    state.num_partitions =
-        static_cast<std::uint32_t>(partition_publishers_.size());
+    to_return = state.publish_promise.get_future();
+    messages_.push_back(std::move(state));
+    if (partition_publishers_.empty()) return to_return;
```

Comment:
nit. add a comment that message will be published whenever partition count is read and publishers are created.